### ISSUE TITLE
Firefox priority handling busts the HTTP/2 server.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8695]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8695|
|Type|defect|
|Created|2016-07-27 08:03:42Z|
|Branch|https://github.com/twisted/twisted/tree/8695-lukasa-firefoxprioritybam|

The way Firefox handles its blast of initial PRIORITY frames busts the Twisted HTTP/2 server:

```
  File "/usr/local/site-packages/twisted/internet/base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/usr/local/site-packages/twisted/web/_http2.py", line 351, in _sendPrioritisedData
    frameData = self._outboundStreamQueues[stream].popleft()
exceptions.IndexError: pop from an empty deque
```

This is almost certainly the result of the stream priorities not being correctly blocked.

<details><summary>Searchable metadata</summary>

```
trac-id__8695 8695
type__defect defect
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__8695_lukasa_firefoxprioritybam 8695-lukasa-firefoxprioritybam
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1469606622912545 1469606622912545
changetime__1469797247720449 1469797247720449
version__None None
owner__Lukasa Lukasa

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

In changeset 7af5a7dc8b88157cda2cfb3ad2d4ef58d0b386ce
```
#!CommitTicketReference repository="" revision="7af5a7dc8b88157cda2cfb3ad2d4ef58d0b386ce"
Failing test for #8695
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_2"></a>|@Lukasa commented|
|-|-|

In changeset ea5ee5efe34b6e408be39443b1d45649ce393da0
```
#!CommitTicketReference repository="" revision="ea5ee5efe34b6e408be39443b1d45649ce393da0"
Add topfile for #8695.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_3"></a>|@Lukasa commented|
|-|-|

See https://github.com/twisted/twisted/pull/429

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_4"></a>|@Lukasa commented|
|-|-|

In changeset 69c0addd2d7281c3a8cc5a6aea1b8ac4cf3248c1
```
#!CommitTicketReference repository="" revision="69c0addd2d7281c3a8cc5a6aea1b8ac4cf3248c1"
Failing test for #8695
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_5"></a>|@Lukasa commented|
|-|-|

In changeset b56d822828d787756f5d1cbc8205d3a6d4d27ce2
```
#!CommitTicketReference repository="" revision="b56d822828d787756f5d1cbc8205d3a6d4d27ce2"
Add topfile for #8695.
```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @Lukasa|
|-|-|

Thanks. Changes look good.

Feel free to merge it as it is now.

After reading the RFC, reading your comments and reading the implementation of buildRequestFrames, the changes look good :)


I would prefer to have better DSL helper methods to make the code easier to read so that we would not need the current comments to explain what the code does... but rather have a code which speaks for itself.

I would prefer to have a helper method which can create just the header frame, rather than forcing the usage of current helper which also created a data/payload frame, which later is ignored in the test.

Check check my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|GitHub set status to `closed`|
|-|-|

In changeset 30ef07f4479c68c8f6b09e8630b862d3a319fee2
```
#!CommitTicketReference repository="" revision="30ef07f4479c68c8f6b09e8630b862d3a319fee2"
Merge 8695-lukasa-firefoxprioritybam: Handle WINDOW_UPDATE frames when we have no data to send.

Author: Lukasa
Reviewer: adiroiban
Fixes: #8695
```
